### PR TITLE
CFE-2686: Policy function format() no longer truncates strings lager than 4KiB

### PR DESCRIPTION
**Pull request by larsewi**

### CODE REVIEW

File: tests/acceptance/01_vars/02_functions/format_edge_case.cf [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Test bug fix
```
Comment (olehermanse) [inline tests/acceptance/01_vars/02_functions/format_edge_case.cf:1]: In commit message:

```
Ticket: CFE-2686-3.18.x
```

Please omit the `3-3.18.x` part.

**Comment by olehermanse** [comment]
@nickanderson please review policy file / test.

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5994/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5994/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5994/

File: libpromises/evalfunction.c [hunk 2]
```diff
@@ -5442,9 +5443,10 @@ static FnCallResult FnCallFormat(EvalContext *ctx, ARG_UNUSED const Policy *poli
                     }
                     else if (strrchr(format_piece, 's'))
                     {
-                        snprintf(piece, CF_BUFSIZE, format_piece, data);
-                        BufferAppend(buf, piece, strlen(piece));
-                        // CfOut(OUTPUT_LEVEL_INFORM, "", "format: appending string format piece = '%s' with data '%s'", format_piece, data);
+                        char *formated;
+                        xasprintf(&formated, format_piece, data);
+                        BufferAppend(buf, formated, strlen(formated));
+                        free(formated);
```
Comment (olehermanse) [inline libpromises/evalfunction.c:5444]: `formatted`; double `t`.

File: libpromises/evalfunction.c [hunk 3]
```diff
@@ -5473,9 +5475,11 @@ static FnCallResult FnCallFormat(EvalContext *ctx, ARG_UNUSED const Policy *poli
                         {
                             Writer *w = StringWriter();
                             JsonWriteCompact(w, value);
-                            snprintf(piece, CF_BUFSIZE, format_rewrite, StringWriterData(w));
+                            char *formated;
+                            xasprintf(&formated, format_rewrite, StringWriterData(w));
```
Comment (olehermanse) [inline libpromises/evalfunction.c:5474]: Same as before, `formatted`.

File: libpromises/evalfunction.c [hunk 4]
```diff
@@ -5473,9 +5475,11 @@ static FnCallResult FnCallFormat(EvalContext *ctx, ARG_UNUSED const Policy *poli
                         {
                             Writer *w = StringWriter();
```
Comment (olehermanse) [inline libpromises/evalfunction.c:5471]: Seems like there is a memory leak here, but it was already there. It's missing a `WriterClose()`.

File: libpromises/evalfunction.c [hunk 5]
```diff
@@ -5498,9 +5502,11 @@ static FnCallResult FnCallFormat(EvalContext *ctx, ARG_UNUSED const Policy *poli
                                 }
                                 WriterWrite(w, " }");
 
-                                snprintf(piece, CF_BUFSIZE, format_rewrite, StringWriterData(w));
+                                char *formated;
+                                xasprintf(&formated, format_rewrite, StringWriterData(w));
```
Comment (olehermanse) [inline libpromises/evalfunction.c:5501]: And here, `formatted`.

File: libpromises/evalfunction.c [hunk 6]
```diff
@@ -5498,9 +5502,11 @@ static FnCallResult FnCallFormat(EvalContext *ctx, ARG_UNUSED const Policy *poli
...
                                 WriterClose(w);
-                                BufferAppend(buf, piece, strlen(piece));
+                                BufferAppend(buf, formated, strlen(formated));
+                                free(formated);
```
Comment (olehermanse) [inline libpromises/evalfunction.c:5504]: Again, there should be a `WriterClose()` here, but that's not your fault.

**Review by olehermanse — APPROVED** [review]
Great! Just some smaller comments.

File: libpromises/evalfunction.c (hunk 4 above)
Comment (olehermanse) [reply libpromises/evalfunction.c:5471]: Created separate ticket for this: https://tracker.mender.io/browse/CFE-3528

File: libpromises/evalfunction.c (hunk 6 above)
Comment (olehermanse) [reply libpromises/evalfunction.c:5504]: Created separate ticket for this: https://tracker.mender.io/browse/CFE-3528

File: tests/acceptance/01_vars/02_functions/format_edge_case.cf [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+bundle agent check {
+  methods:
+      "check"  usebundle => dcs_check_state(test,
+                                            "$(this.promise_filename).expected.json",
+                                            $(this.promise_filename));
+
+  reports:
+    "list_str: $(test.list_str)";
+    "container_str: $(test.container_str)";
```
Comment (vpodzime) [inline tests/acceptance/01_vars/02_functions/format_edge_case.cf:31]: Please drop these `reports:` promises or add a `DEBUG::` class guard.

File: libpromises/evalfunction.c [hunk 8]
```diff
@@ -5334,9 +5334,8 @@ static FnCallResult FnCallSort(EvalContext *ctx, ARG_UNUSED const Policy *policy
 
 static FnCallResult FnCallFormat(EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
 {
-    char id[CF_BUFSIZE];
-
-    snprintf(id, CF_BUFSIZE, "built-in FnCall %s-arg", fp->name);
+    char *id;
+    xasprintf(&id, "built-in FnCall %s-arg", fp->name);
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply libpromises/evalfunction.c:5338]: Can't this be just a string literal? I mean, `fp->name` is always `"format"`, right?

File: libpromises/evalfunction.c (hunk 2 above)
Comment (vpodzime) [inline libpromises/evalfunction.c:5444]: There is `BufferAppendF()` which does the same as the functions above.

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise. Thanks for the code quality improvements!

File: libpromises/evalfunction.c (hunk 2 above)
Comment (larsewi) [reply libpromises/evalfunction.c]: Nice @vpodzime !

File: libpromises/evalfunction.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (larsewi) [reply libpromises/evalfunction.c:5338]: Should I leave it or change it @olehermanse @vpodzime ?

[the remark this answers is not in the document]
Comment (olehermanse) [reply libpromises/evalfunction.c:5338]: @larsewi for high-score, I'd say remove this change from this commit and make a separate commit where you remove the `snprintf` and assign a string literal instead.

File: libpromises/evalfunction.c [hunk 9]
```diff
@@ -5334,9 +5334,7 @@ static FnCallResult FnCallSort(EvalContext *ctx, ARG_UNUSED const Policy *policy
 
 static FnCallResult FnCallFormat(EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
 {
-    char id[CF_BUFSIZE];
-
-    snprintf(id, CF_BUFSIZE, "built-in FnCall %s-arg", fp->name);
+    const char *const id = "buildt-in FnCall format-arg";
```
Comment (olehermanse) [inline libpromises/evalfunction.c:5337]: Oops, a `d` snuck in there, should be `built-in`.

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5996/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5996/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5996/

**Review by olehermanse — APPROVED** [review]
Looks great now, awaiting CI.

**Comment by olehermanse** [comment]
Cherry-picks:
https://github.com/cfengine/core/pull/4431
https://github.com/cfengine/core/pull/4432

**Pull request merged**